enhancements/baremetal/baremetal-poison-pill-remediation.md

```diff
@@ -0,0 +1,189 @@
+and restore cluster capacity (where possible).
+
+This proposal describes an alternate mechanism for a node in a cluster
+to detect its health status and take actions to remediate itself in case of a failure. While not all remediation events
+can result in the node returning to a healthy state, the proposal does allow surviving parts of the cluster to assume
+the node has reached a safe state so that its workloads can be automatically recovered.
+
+
+This work can also be useful for clusters **with** BMC credentials.
+If there’s a network outage between the node running CAPBM (which expects to use BMC commands like power-off etc.) and the
+unhealthy node, the power-cycle attempt cannot
+succeed. Self health checks and remediation can resolve such cases, however that work/scenario is out of scope for
+this proposal.
+
+## Motivation
+
+Some clusters don’t have BMC credentials and we still want to have auto remediation for failing nodes.
+This proposal allows:
+1. To allow stateful workload running on the unhealthy nodes to be rescheduled after finite time on other healthy nodes.
+1. To restore compute capacity by remediating the unhealthy node
+
+### Goals
+
+* To allow remediation of bare metal nodes without BMC credentials
+* To utilize the MHC logic for deciding if a node is unhealthy and whether remediation is appropriate
+* To allow the healthy majority to recover workloads after a pre-determined and finite interval
+* To allow a node to take actions to remediate itself in case of a failure
+* To allow a node to detect its health status when master<->node communication is partially failed
+* To avoid false positives caused by failure of the control plane
+
+### Non-Goals
+
+* Self healing in combination with existing reprovisioning or reboot based remediation mechanisms in the Machine Health Check controller
+* Recovering from all types of failure
+* Having a distributed mesh of health checks where each node is checking health for its peers
+* Creation of fake Machine objects for UPI clusters
+
+## Proposal
+
+Utilizing a DaemonSet to have a privileged critical pod running on each node. This pod will periodically check the
+health status of the node it’s running on, and tickle a [watchdog device](https://en.wikipedia.org/wiki/Watchdog_timer).
+
+This will be performed by looking for the `host.metal3.io/external-remediation` annotation added by MHC to the unhealthy
+Machine CR.
+
+If the unhealthy annotation was found, the node will remediate itself by using the watchdog device to
+trigger a reset of the node. Watchdog devices, especially the hardware kind, are considered more reliable and predictable than the
+reboot command.
+
+If Machine CR is not accessible in 3 consecutive attempts, it needs external assistance to detect its
+health status.
+
+Failure to read Machine CR could stem from one of the following:
+1. The node is unhealthy and therefore can’t read the Machine CR (e.g. local network failure)
+1. There’s an etcd failure
```

Comment:
do we risk losing quorum by rebooting nodes like proposed here? cc @hexfusion

Reply:
We can reboot only if etcd-quorum-guard PDB allows it, but the watchdog doesn't respect PDBs obviously.

Maybe at this stage it would be simpler to to install only on workers (and lose the opportunity to remediate masters). 
We can revisit this once we gain more experience and confidence.

cc @beekhof